core/src/banking_stage/transaction_scheduler/scheduler_controller.rs

```diff
@@ -0,0 +1,618 @@
+        scheduler: PrioGraphScheduler,
+    ) -> Self {
+        Self {
+            decision_maker,
+            packet_receiver: packet_deserializer,
+            bank_forks,
+            transaction_id_generator: TransactionIdGenerator::default(),
+            container: TransactionStateContainer::with_capacity(TOTAL_BUFFERED_PACKETS),
+            scheduler,
+        }
+    }
+
+    pub fn run(mut self) -> Result<(), SchedulerError> {
+        loop {
+            let decision = self.decision_maker.make_consume_or_forward_decision();
+            self.process_transactions(&decision)?;
+            self.scheduler.receive_completed(&mut self.container)?;
+            if self.receive_packets(&decision) {
+                break;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Process packets based on decision.
+    fn process_transactions(
+        &mut self,
+        decision: &BufferedPacketsDecision,
+    ) -> Result<(), SchedulerError> {
+        match decision {
+            BufferedPacketsDecision::Consume(_bank_start) => {
+                let _num_scheduled = self.scheduler.schedule(&mut self.container)?;
+            }
+            BufferedPacketsDecision::Forward => {
+                self.clear_container();
+            }
+            BufferedPacketsDecision::ForwardAndHold | BufferedPacketsDecision::Hold => {}
+        }
+
+        Ok(())
+    }
+
+    /// Clears the transaction state container.
+    /// This only clears pending transactions, and does **not** clear in-flight transactions.
+    fn clear_container(&mut self) {
+        while let Some(id) = self.container.pop() {
+            self.container.remove_by_id(&id.id);
+        }
+    }
+
+    /// Returns whether the packet receiver was disconnected.
+    fn receive_packets(&mut self, decision: &BufferedPacketsDecision) -> bool {
+        let remaining_queue_capacity = self.container.remaining_queue_capacity();
+
+        let (recv_timeout, should_buffer) = match decision {
+            BufferedPacketsDecision::Consume(_) => (
+                if self.container.is_empty() {
+                    Duration::from_millis(100)
+                } else {
+                    Duration::from_millis(0)
+                },
+                true,
+            ),
+            BufferedPacketsDecision::Forward => (Duration::from_millis(100), false),
+            BufferedPacketsDecision::ForwardAndHold | BufferedPacketsDecision::Hold => {
+                (Duration::from_millis(100), true)
+            }
+        };
+
+        let received_packet_results = self
+            .packet_receiver
+            .receive_packets(recv_timeout, remaining_queue_capacity);
+
+        match (received_packet_results, should_buffer) {
+            (Ok(receive_packet_results), true) => {
+                self.buffer_packets(receive_packet_results.deserialized_packets)
+            }
+            (Ok(receive_packet_results), false) => drop(receive_packet_results),
```

Comment:
946afd1e074ea3cfa1f41ac3d26077f83fca2c2b

Added a long comment here about the two. Then moved it to the decision making location: c165557483004e9229d06e756f867ca698fa39d6

I've been running experiments without forwarding which have shown no negative impact (both without receiving forwarded or forwarding). Having some discussions with others about removing it entirely still, but for the time being I am moving forward with the new `BankingStage` without forwarding. At least until we see evidence or reason to actually add it in.

Since I already have written the `ForwardWorker` stuff, it's relatively easy to add forwarding back in at this point.
For now, to clarify on this comment directly, since we are not forwarding we can just `drop` these packets so we don't waste time sanitizing and buffering.

Reply:
Sorry for the delay. Wanted to think over "no forwarding" again, since it feels to me as a big change by itself. It doesn't really belong to this PR, since it is being discussed separately. Happy to leave it as is for now.